File: app/src/main/java/antoninovitale/dropcodechallenge/details/BeerDetailsPresenterImpl.java
package antoninovitale.dropcodechallenge.details;

import java.util.List;

import antoninovitale.dropcodechallenge.api.model.Beer;
import antoninovitale.dropcodechallenge.details.section.model.HeaderSectionModel;
import antoninovitale.dropcodechallenge.details.section.model.IngredientSectionModel;
import antoninovitale.dropcodechallenge.details.section.model.MethodSectionModel;
import antoninovitale.dropcodechallenge.details.section.model.Status;

/**
 * Created by antoninovitale on 31/08/2017.
 */
public class BeerDetailsPresenterImpl implements BeerDetailsPresenter, BeerDetailsInteractor
        .OnStatusCheckListener, BeerDetailsInteractor.OnSectionSetupListener {
    private final BeerDetailsInteractor interactor;

    private BeerDetailsView view;

    public BeerDetailsPresenterImpl() {
        this.interactor = new BeerDetailsInteractorImpl();
    }

    @Override
    public void setView(BeerDetailsView view) {
        this.view = view;
    }

    @Override
    public void onChanged(Beer beer) {
        interactor.setupBeer(beer, this);
    }

    @Override
    public void onStatusClick(int position) {
        interactor.checkIngredientStatus(position, this);
    }

    @Override
    public void onMethodStatusClick(int position) {
        interactor.checkMethodStatus(position, this);
    }

    @Override
    public void onMethodEnd(int position) {
        interactor.setMethodDone(position, this);
    }

    @Override
    public void onMethodTimeElapsed(int position, long millis) {
        interactor.setMethodTimeElapsed(position, millis);
    }

    @Override
    public void onMaltDone(int position) {
        view.notifyMaltDone(position);
    }

    @Override
    public void onHopDone(int position) {
        view.notifyHopDone(position);
    }

    @Override
    public void onMethodDone(int position) {
        view.notifyMethodStatusChanged(position, Status.DONE);
    }

    @Override
    public void onMethodStatusChanged(int position, Status status) {
        view.notifyMethodStatusChanged(position, status);
    }

    @Override
    public void onHeaderSet(HeaderSectionModel model) {
        view.setupHeaderSection(model);
    }

    @Override
    public void onMaltsSet(List<IngredientSectionModel> malts) {
        view.setupMaltsSection(malts);
    }

    @Override
    public void onHopsSet(List<IngredientSectionModel> hops) {
        view.setupHopsSection(hops);
    }

    @Override
    public void onMethodsSet(List<MethodSectionModel> methods) {
        view.setupMethodSection(methods);
    }

}

File: app/src/main/java/antoninovitale/punkapi/app/details/BeerDetailsPresenterImpl.kt
package antoninovitale.punkapi.app.details

import antoninovitale.punkapi.app.api.model.Beer
import antoninovitale.punkapi.app.details.BeerDetailsInteractor.OnSectionSetupListener
import antoninovitale.punkapi.app.details.BeerDetailsInteractor.OnStatusCheckListener
import antoninovitale.punkapi.app.details.section.model.HeaderSectionModel
import antoninovitale.punkapi.app.details.section.model.IngredientSectionModel
import antoninovitale.punkapi.app.details.section.model.MethodSectionModel
import antoninovitale.punkapi.app.details.section.model.Status
import antoninovitale.punkapi.app.details.section.model.Status.DONE

/**
 * Created by antoninovitale on 31/08/2017.
 */
class BeerDetailsPresenterImpl : BeerDetailsPresenter, OnStatusCheckListener,
        OnSectionSetupListener {
    private val interactor: BeerDetailsInteractor
    private var view: BeerDetailsView? = null

    init {
        interactor = BeerDetailsInteractorImpl()
    }

    override fun setView(view: BeerDetailsView?) {
        this.view = view
    }

    override fun onChanged(beer: Beer) {
        view?.clearSections()
        interactor.setupBeer(beer, this)
    }

    override fun onStatusClick(position: Int) {
        interactor.checkIngredientStatus(position, this)
    }

    override fun onMethodStatusClick(position: Int) {
        interactor.checkMethodStatus(position, this)
    }

    override fun onMethodEnd(position: Int) {
        interactor.setMethodDone(position, this)
    }

    override fun onMethodTimeElapsed(position: Int, millis: Long) {
        interactor.setMethodTimeElapsed(position, millis)
    }

    override fun onMaltDone(positionInSection: Int, globalPosition: Int) {
        view?.notifyMaltDone(positionInSection, globalPosition)
    }

    override fun onHopDone(positionInSection: Int, globalPosition: Int) {
        view?.notifyHopDone(positionInSection, globalPosition)
    }

    override fun onMethodDone(positionInSection: Int, globalPosition: Int) {
        view?.notifyMethodStatusChanged(positionInSection, globalPosition, DONE)
    }

    override fun onMethodStatusChanged(positionInSection: Int, globalPosition: Int, status: Status) {
        view?.notifyMethodStatusChanged(positionInSection, globalPosition, status)
    }

    override fun onHeaderSet(model: HeaderSectionModel) {
        view?.setupHeaderSection(model)
    }

    override fun onMaltsSet(malts: List<IngredientSectionModel>) {
        view?.setupMaltsSection(malts)
    }

    override fun onHopsSet(hops: List<IngredientSectionModel>) {
        view?.setupHopsSection(hops)
    }

    override fun onMethodsSet(methods: List<MethodSectionModel>) {
        view?.setupMethodSection(methods)
    }
}